# Introduction to ILP and LP Relexation (Lecture 15)

### Author: Mengshen Yun (myun7)                                                              

Linear programming is a method to achieve the best outcome such as maximum profit or lowest cost in a mathematiacl model whose requirements are represented by linear relationships. It is widely adopted to solve a variety of mathematic and engineering problems, where the functions are subject to a set of linear inequality or equality constraints.

### Real-world problems solved by linear/integer linear program

 To illustrate the use of linear program, a real-life toy example with only three variables is presented in this section. 

For example, in a candy shop, there are three types of candies produced by the workers: Candy A($3.75), Candy B($4.25), and Candy C($1.25). There is 2 candy machines, 5 workers and 2 retailers. Let's assume the machines work 24 hours per day, the workers work 7 hours per day, and the retailers work 6 hours per day.

The time needed to produce and sell each type of candies is different:

Machine time: Candy A--2.75 mins; Candy B--3.25mins; Candy C--2mins

Worker time: Candy A--3.5 mins; Candy B--3.75mins; Candy C--1mins

Retailer time: Candy A--1.5 mins; Candy B--3mins; Candy C--2mins

The function for profit can be modelled as:

$\\ Profit = 1.5A + 3B + 2C$

The constraints are:

$A\geq 0 \,\,\, B\geq 0 \,\,\,C\geq 0 \,\,\,\,\,\, \forall A,B,C \in \mathbb{Z} \\
2.75A+3.25B+2C \leq 1440 \\
3.5A+3.75B+C \leq 420\\
1.5A+3B+2C \leq 360 $

In [31]:
import pulp
import time

In [32]:
# Instantiate our problem class
model = pulp.LpProblem("Profit maximizing problem", pulp.LpMaximize)

In [33]:
# Constraints on variables: non-negative and integers
A = pulp.LpVariable('CandyA', lowBound=0, cat='Integer')
B = pulp.LpVariable('CandyB', lowBound=0, cat='Integer')
C = pulp.LpVariable('CandyC', lowBound=0, cat='Integer')

In [34]:
# Objective function
model += 3.75*A + 4.25*B + 1.25*C, "Profit"
# Constraints
model += 2.75 * A + 3.25 * B + 2*C <= 1440
model += 3.5 * A + 3.75 * B + 1*C <= 420
model += 1.5 * A + 3 * B + 2*C <= 360

In [35]:
# Solve our problem
start = time.time()
model.solve()
end = time.time()
pulp.LpStatus[model.status]
print("Optimal solution time =", end - start)

Optimal solution time = 0.09774041175842285


In [36]:
# Print our decision variable values
print( "Production of Candy A = ",format(A.varValue))
print ("Production of Candy B = ",format(B.varValue))
print( "Production of Candy C = ",format(C.varValue))
# Print our objective function value
print ("Maximized profit:",pulp.value(model.objective))

Production of Candy A =  1.0
Production of Candy B =  106.0
Production of Candy C =  19.0
Maximized profit: 478.0


The result shows that the store needs to produce 1 piece of CandyA, 106 pieces of Candy B, and 19 pieces of Candy C in order to maximize the profit. By following this pattern, the maximum profit that the store can get per day is 478.3 dollars!

Let's explore what will happen if we remove one of the constraints (linear program relaxation). In the following model, variables are no longer required to be integers. Instead, they can be any values that are greater than or equal to zero.

In [37]:
# Instantiate our problem class
model = pulp.LpProblem("Profit maximizing problem", pulp.LpMaximize)
A = pulp.LpVariable('CandyA', lowBound=0)
B = pulp.LpVariable('CandyB', lowBound=0)
C = pulp.LpVariable('CandyC', lowBound=0)
# Objective function
model += 3.75*A + 4.25*B + 1.25*C, "Profit"

# Constraints
model += 2.75 * A + 3.25 * B + 2*C <= 1440
model += 3.5 * A + 3.75 * B + 1*C <= 420
model += 1.5 * A + 3 * B + 2*C <= 360

# Solve our problem
start = time.time()
model.solve()
end = time.time()
pulp.LpStatus[model.status]
print("Optimum solution time =", end - start)
# Print our decision variable values
print( "Production of Candy A = ",format(A.varValue))
print ("Production of Candy B = ",format(B.varValue))
print( "Production of Candy C = ",format(C.varValue))
# Print our objective function value
print ("Maximized profit:",pulp.value(model.objective))

Optimum solution time = 0.039893150329589844
Production of Candy A =  0.0
Production of Candy B =  106.66667
Production of Candy C =  20.0
Maximized profit: 478.3333475


This time, the result shows that the store needs to produce 0 piece of CandyA, 106.66 pieces of Candy B, and 20 pieces of Candy C in order to maximize the profit. By following this pattern, the maximum profit that the store can get per day is still 478.3 dollars.

Here, we may notice that the solution of this program is similar to the program above. Though the solutions of the variables are not integers anymore, the values are still pretty close. However, the time the solve took for finding the optimum solutions of two program is very different!

 It took 0.0977 ms to solve the integer linear program, but only took 0.0398 ms to solve the linear program with relaxation.

### Image segmentation with ILP and LP relaxation

After we have walked through the toy example together, we should be ready for something more complex and exciting. Besides real-world resourcing problems, integer linear programs are also widely used in the field of machine learning for solving rather complex problems such as structured prediction models. In this section, we will implement a structured prediction model to perform image segmentation.